src/main/java/com/linkedin/kafka/clients/consumer/ConsumerRecordsProcessor.java

```diff
@@ -304,63 +284,66 @@ public void clear(TopicPartition tp) {
+    MessageAssembler.AssembleResult assembledResult = _messageAssembler.assemble(topicPartition, srcRecord.offset(), srcRecord);
+
+    LOG.trace("Got message {} from partition {}", srcRecord.offset(), topicPartition);
+
+    long safeOffset = Math.min(srcRecord.offset() + 1, _messageAssembler.safeOffset(topicPartition));
+    if (shouldSkip(topicPartition, srcRecord.offset())) {
+      //The message had already been delivered to the consumer when it committed and it should not be seen again.
+      LOG.trace("Skipping message {} from partition {} because its offset is smaller than the high watermark",
+          srcRecord.offset(), topicPartition);
+      _deliveredMessageOffsetTracker.track(topicPartition, srcRecord.offset(), safeOffset, srcRecord.offset(), false);
+      return null;
     }
-    return handledRecord;
-  }
 
-  private byte[] parseAndMaybeTrackRecord(TopicPartition tp, long messageOffset, byte[] bytes) {
-    MessageAssembler.AssembleResult assembledResult = _messageAssembler.assemble(tp, messageOffset, bytes);
-    if (assembledResult.messageBytes() != INCOMPLETE_RESULT) {
-      LOG.trace("Got message {} from partition {}", messageOffset, tp);
-      boolean shouldSkip = shouldSkip(tp, messageOffset);
-      // The safe offset is the smaller one of the current message offset + 1 and current safe offset.
-      long safeOffset = Math.min(messageOffset + 1, _messageAssembler.safeOffset(tp));
-      _deliveredMessageOffsetTracker.track(tp,
-                                           messageOffset,
-                                           safeOffset,
-                                           assembledResult.messageStartingOffset(),
-                                           !shouldSkip);
-      // We skip the messages whose offset is smaller than the high watermark.
-      if (shouldSkip) {
-        LOG.trace("Skipping message {} from partition {} because its offset is smaller than the high watermark",
-                  messageOffset, tp);
-        return INCOMPLETE_RESULT;
-      } else {
-        return assembledResult.messageBytes();
-      }
-    } else {
-      _deliveredMessageOffsetTracker.addNonMessageOffset(tp, messageOffset);
-      return INCOMPLETE_RESULT;
+    if (assembledResult == null) {
+      //Not a large message segment
+      _deliveredMessageOffsetTracker.track(topicPartition, srcRecord.offset(), safeOffset, srcRecord.offset(), true);
+      return srcRecord;
+    }
+
+    if (assembledResult.messageBytes() == null) {
+      //Not a complete, large message
+      _deliveredMessageOffsetTracker.addNonMessageOffset(topicPartition, srcRecord.offset());
+      return null;
     }
+
+    // Completed, large message value
+    _deliveredMessageOffsetTracker.track(topicPartition, srcRecord.offset(), safeOffset, assembledResult.messageStartingOffset(),
+        true);
+
+    int serializedKeySize = assembledResult.isOriginalKeyNull() ? 0 : srcRecord.key().length;
```

Comment:
Avoiding this double checking will actually result in more lines of code.  Also both of these are conditional assignment statements which should get compiled to CMOV instructions which are easily pipelined since the CPU does not need to do any branch prediction on CMOV.